### PR TITLE
(Re)Moving code that is out of its place

### PR DESCRIPTION
**Pull request by raonismaneoto**
The following issues have been addressed in this pr:

- Removal of useless code, like addUserData();

- Extraction of hard coded scripts into script file;

- Extraction of common code to common module;

- Some general fixes related to Exceptions.

### CODE REVIEW

File: bin/agent-scripts/generateSshKeyPair [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
```
[the remark this answers is not in the document]
Comment (raonismaneoto) [reply bin/agent-scripts/generateSshKeyPair]: For now, only DFNS needs this script and that's why I've moved it to dfns folder. Later on, when we remove the pre installed script dependency from the deploy a common folder might be necessary.

File: src/main/java/cloud/fogbow/fns/core/serviceconnector/ServiceConnectorFactory.java [hunk 2]
```diff
@@ -1,11 +1,11 @@
 package cloud.fogbow.fns.core.serviceconnector;
 
+import cloud.fogbow.common.util.BashScriptRunner;
```
Comment (pauloewerton) [inline src/main/java/cloud/fogbow/fns/core/serviceconnector/ServiceConnectorFactory.java]: is this import being used?

File: src/main/java/cloud/fogbow/fns/core/serviceconnector/DfnsServiceConnector.java [hunk 3]
```diff
@@ -95,29 +98,15 @@ public final UserData getTunnelCreationInitScript(String federatedIp, FederatedC
...
-
-        String[] catCommand2 = {"cat", keyName + ".pub"};
-        BashScriptRunner.Output catResult2 = runner.runtimeRun(catCommand2);
-        String publicKey = catResult2.getContent();
-
-        String[] removeKeysCommand = {"rm", keyName, keyName + ".pub"};
-        BashScriptRunner.Output removeKeysCommandResult = runner.runtimeRun(removeKeysCommand);
-
-        return new String[]{publicKey, privateKey};
+        return new String[]{createCommandResult.getContent().split(KEY_PAIR_SEPARATOR)[0], createCommandResult.getContent().split(KEY_PAIR_SEPARATOR)[1]};
```
Comment (pauloewerton) [inline src/main/java/cloud/fogbow/fns/core/serviceconnector/DfnsServiceConnector.java]: nit: the indexes could be put on constants as to improve readability

File: src/main/java/cloud/fogbow/fns/core/serviceconnector/DfnsServiceConnector.java [hunk 4]
```diff
@@ -95,29 +98,15 @@ public final UserData getTunnelCreationInitScript(String federatedIp, FederatedC
...
-    public String[] generateSshKeyPair() throws UnexpectedException {
+    protected String[] generateSshKeyPair() throws UnexpectedException {
         BashScriptRunner runner = new BashScriptRunner();
         String keyName = String.valueOf(UUID.randomUUID());
 
-        String[] createCommand = {"ssh-keygen", "-t", "rsa", "-b", "1024", "-f", keyName, "-q", "-N", ""};
-        BashScriptRunner.Output createCommandResult = runner.runtimeRun(createCommand);
+        // The key's size is passed as parameter and set to 1024 to keep the key small.
+        String[] genCommand = {"bash", GEN_KEY_PAIR_SCRIPT_PATH, keyName, KEY_SIZE};
+        BashScriptRunner.Output createCommandResult = runner.runtimeRun(genCommand);
```
Comment (pauloewerton) [inline src/main/java/cloud/fogbow/fns/core/serviceconnector/DfnsServiceConnector.java]: did your tests run ok using the bashscriptrunner with an external script?

File: src/main/java/cloud/fogbow/fns/core/serviceconnector/DfnsServiceConnector.java [hunk 5]
```diff
@@ -35,6 +35,9 @@
     public static final String VLAN_ID_ENDPOINT = "/vlanId";
     public static final int PUBLIC_KEY_INDEX = 0;
     public static final int PRIVATE_KEY_INDEX = 1;
+    private final String GEN_KEY_PAIR_SCRIPT_PATH = Paths.get("").toAbsolutePath().toString() + "/bin/agent-scripts/dfns/generateSshKeyPair";
```
Comment (pauloewerton) [inline src/main/java/cloud/fogbow/fns/core/serviceconnector/DfnsServiceConnector.java]: nit: maybe the script path should be in its own constant?

**Comment by pauloewerton** [comment]
thank you for the PR @raonismaneoto . i've left a few nits and questions inline.

File: src/main/java/cloud/fogbow/fns/core/serviceconnector/DfnsServiceConnector.java (hunk 4 above)
Comment (raonismaneoto) [reply src/main/java/cloud/fogbow/fns/core/serviceconnector/DfnsServiceConnector.java]: yeah

File: src/main/java/cloud/fogbow/fns/core/serviceconnector/ServiceConnectorFactory.java (hunk 2 above)
Comment (raonismaneoto) [reply src/main/java/cloud/fogbow/fns/core/serviceconnector/ServiceConnectorFactory.java]: yes, it is

**Comment by raonismaneoto** [comment]
Thank you @pauloewerton @fubica for your comments. Hope I've addressed them correctly. In any case, I'm waiting for your next reviews :)

File: src/main/java/cloud/fogbow/fns/core/serviceconnector/ServiceConnectorFactory.java (hunk 2 above)
Comment (pauloewerton) [reply src/main/java/cloud/fogbow/fns/core/serviceconnector/ServiceConnectorFactory.java]: i missed the constructor on line 30, sorry

**Pull request merged**